### ISSUE TITLE
Wrong Result When Adding `UNWIND`

### ISSUE THREAD

**Issue by CelineWuest** [neo4j/neo4j#13130]
I found a discrepancy when running two semantically equivalent queries against an empty database:

---
1:
```cypher
CALL { OPTIONAL MATCH () CREATE () } MATCH (x) WHERE false UNWIND [0] AS n0 DELETE x
-
RETURN COUNT { () } AS n
```
Produces:
```
╒═══╕
│"n"│
╞═══╡
│1  │
└───┘
```

---
2:
```cypher
CALL { OPTIONAL MATCH () CREATE () } UNWIND [0] AS n1 MATCH (x) WHERE false UNWIND [0] AS n0 DELETE x
-
RETURN COUNT { () }
```
Produces:
```
╒═══╕
│"n"│
╞═══╡
│0  │
└───┘
```
---

I believe the queries mentioned above are semantically equivalent and thus their result should be the same.
The only difference between the two is the inclusion of the first `UNWIND` clause in the second query. As it only iterates over an array of size one, it shouldn't have an effect on the query's outcome.

I encountered this issue when testing queries against the **neo4j:5.6.0** Docker container.

### Steps to reproduce
1.
	<details>

	<summary>Run the Docker Container</summary>

	```
	docker run \
	    -p 7474:7474 -p 7687:7687 \
	    --name neo4j-apoc \
	    -e apoc.export.file.enabled=true \
	    -e apoc.import.file.enabled=true \
	    -e apoc.import.file.use_neo4j_config=true \
	    -e NEO4J_PLUGINS=\[\"apoc\"\] \
	    -e NEO4J_AUTH=none
	    --rm \
	    neo4j:5.6.0
	```

	</details>

1. Navigate to the browser interface on [localhost:7474](http://localhost:7474)

1. Run the aforementioned queries and observe that their results differ.

### Expected behavior
The queries should result in the same behavior

### Actual behavior
The first query returns `1`, whereas the second query returns `0`.

**Comment by AzuObs** [comment]
Hi Dominic,

Thank you very much for your report. This is very odd, and was not happening with the 5.6.0 enterprise version of the software. I have submitted the bug to a team for further investigation ([private link](https://trello.com/c/G11VW1u1)).

Hope you have a nice day,

**Comment by CelineWuest** [comment]
Thanks for looking into it @AzuObs!
I noticed that it doesn't trigger in the sandbox too, which I thought was weird. Additionally, when following the steps in **Steps to reproduce** section but changing the tag to 5.4, the bug doesn't trigger. Maybe this info could also be helpful?

**Comment by AzuObs** [comment]
Thanks Dominic, I think it will help. By the way, although I couldn't reproduce on enterprise, I could definitely reproduce on community 5.6.0 

I think that's why it is happening in some places an not others, probably the enterprise code does this correctly but the community version has a bug.

**Comment by InverseFalcon** [comment]
To note, one major difference between the two is that community can only use SLOTTED runtime, enterprise would likely be using PIPELINED runtime. That suggests the issue may be a SLOTTED runtime bug, and gives us some more options for trying to reproduce this.

**Comment by sherfert** [comment]
Hi. We implemented a fix which will be included in 5.9 and 4.4.21.
